Adding Messiaen's 7 "Modes of Limited Transposition"

rowild opened the pull request:
I would like to suggest/discuss the adding of Messiaen's 7 "Modes Of Limited Transposition". The addition consists partly of adding a new name to an existing scale, but also adding new scales.

As for the naming I am unsure what would be the best. Currently "messiaen's mode of limited transposition #x" is used, which is exact, but also very long and uses an apostrophe, which might cause troubles. Personally, I would be fine with something like "messiaen #x" or "mode of limited transposition #x" (which would at least get rid of the potential apostrophe problem).

Source:
https://en.wikipedia.org/wiki/Mode_of_limited_transposition
https://es.wikipedia.org/wiki/Modos_de_transposici%C3%B3n_limitada
https://de.wikipedia.org/wiki/Modi_mit_begrenzten_Transpositionsm%C3%B6glichkeiten

Muchas gracias, Daniel!

danigb commented:
Thanks a lot for the PR 🎉 I'd like to adding them, of course. 

Before merging, I think the PR needs to:

- Fix the tests. At least, there's a test to check the number of scales. Just run `npx jest` at root folder
- I like a lot you paste the wikipedia reference to them. I think `packages/scale-type/README.md` would benefit of having those links. I'd add a new `## References` section at the bottom of the README with this link (and maybe others you consider can be a good reference for the included scales)

I liked you moved the blues scale to it's position 👌 

Regarding the name, I like something like "Messiaen's first mode", "Messiaen's second mode" but I don't know if it's accurate enough

Danke!! 🙏

rowild commented:
OK, this will take a while, because the weakest of all my weak spots is testing! Totally no practical experience at all... but I want to give it a try!

I was not aware that i moved a scale. Admittedly, though, I do not recognise a structure there aside from the very welcome 5-note-, 6-note-scales etc grouping. If that is a topic, it should probably be done in a separate PR.

@ Naming: 
The apostrophe is not a problem? But lowercase still is the rule to abide by, right?

I'd rather would like to have a naming, that "orders itself". Putting "first" and "second" etc in the middle would not really help, when there should be need to sort the modes, while sth like "messiaen #1", "messiaen #2" etc would.

Then again I am of course aware that you put a lot of effort into the naming of all those scales, and they turned out really nicely. So I do know that my "wish" does not really go along well with your style.

However, I ask you for your opinion on that once more, and then maybe you can just decide and we get it done. There is always the possibility to override sth in one's own code.

gracias otra vez y hasta la proxima!

rowild commented:
I need your help: I have no idea, how to solve this - any instructions?

```
Test suite failed to run

    packages/key/test.ts:2:30 - error TS2307: Cannot find module '@tonaljs/scale'.
```

I saw that "scale" is marked "deprecated" in the export of ´scale/index.ts` - is that of any relevance?

danigb commented:
Hi, 

No worries. Testing is not so hard. First of all, I think you need to run at root folder (you need yarn. npm doesn't work):
```
yarn install
yarn build
```

Take a look to CONTRIBUTING.md. Hope it helps. Good luck and don't hesitate to ask

#### Naming

> The apostrophe is not a problem? 

Not in my opinion

> But lowercase still is the rule to abide by, right?

Not necessarily... 🤔 but not sure about that

> I'd rather would like to have a naming, that "orders itself".

Good point!

It think `"Messiaen's #1"` (or `"messiaen's #1"`) would work for me... unless he has already other different scales with his name 😂 Any news about this?

What do you think?

rowild commented:
I think sticking with lowercase would fit well with all your current naming ("locrian", "blues"... everything is lowercase, I would vote for keeping it that way).

Haha - more "Messiaen"-named scales? I hope not! At least I am not aware of any... But you mention a very important fact: "Messiaen" is actually not part of those scales names. i added the name to make them more recognisable within your library, which is "multi culti" so-to-say, Indian next to Jazz next to Classic next to Church Modes... Which is awesome! I simply thought it might help those, who read the code, to quickly grasp where those mode come from. Simply using "mode of limited transposition #1" is very technical – no flair of style (like "blues scale") or history (like "aeolian") or culture (like any of those Indian and Gamelan scales) - while "messiaen's mode #1" has at least a bit of personality ringing along with it.

So - if there is no vote against it, I will name the scales "messiaen's mode #1" etc, a compromise between length and accuracy (after all, "mode" is a keyword).

@testing

So finally I end up using yarn. Sth I tried to avoid... :-) But hey, it's fast! and colorful!
The tests work fine now. The one, that failed, now expects 93 instead of 88 scale types, so I changed that in the `scale-type/test.js` file.
`yarn build` also works.

The next step would be to upload the changes to my forked repo, right? When doing  "git status", I see that there is a new folder, "packages/tonal/browser/**" that git picks up, too. Do I commit this folder, too?

Otra vez muchas gracias!

danigb commented:
Hi!

> I think sticking with lowercase would fit well

👌 

> more "Messiaen"-named scales? I hope not! At least I am not aware of any

Excellent!!

> your library, which is "multi culti" so-to-say,

😂 exactly, a cut and paste from different sources

> So - if there is no vote against it, I will name the scales "messiaen's mode #1" etc, a compromise between length and accuracy (after all, "mode" is a keyword).

Agree! 🎉 

> So finally I end up using yarn ... 

Yes. This library uses a feature only present in yarn (https://classic.yarnpkg.com/en/docs/workspaces/).

> woks!

🎉  Good news! Thanks for updating the tests 👍 

> there is a new folder, "packages/tonal/browser/**"

Yes, this is the browser file that is also built to add your brand new scales!

rowild commented:
I think I added everything we wanted, but I have a strong feeling that this a not a very clean PR. especially yesterday's PR with the quotes problem with the aeolian scale name might cause troubles (I first added them here, too, then removed then again... I simply don't know how to undo something, once I changed a bit of content).

Maybe you can merge it anyway, otherwise please let me know and I provide a clean PR.

Gracias como siempre por tu ayuda, tu paciencia y simplemente por todo, Daniel!

danigb commented on packages/scale-type/data.ts:
```diff
@@ -31,14 +31,15 @@ const SCALES: string[][] = [
   ["1P 2m 3M 5d 6m 7m", "mystery #1"],
   ["1P 2m 3M 4P 5A 6M", "six tone symmetric"],
-  ["1P 2M 3M 4A 5A 7m", "whole tone"],
+  ["1P 2M 3M 4A 5A 7m", "whole tone", "messiaen's mode of limited transposition #1"],
```
I thought we decided the name would me "messiaen's mode #1" 😂

danigb commented:
It not a problem how many commits you do: when merged the will converted into one. So do whatever you need to make it poilshed 👍 

The only "problem" I see is the name of the scales... a little bit too long for my taste :-)

Thanks!!

rowild commented:
damn it - i forgot those!!! Sorry - will be fixed in a moment...

rowild commented:
OK - this should do it! Is that ok with you? Can you merge it?
I'd love to know, which problems occur and where I must pay more attention in future!

Have a good day!

danigb commented:
Awesome! 🎉  Thanks!
